### PR TITLE
Bug fix /transaction/cost api route

### PR DESCRIPTION
**Pull request by miiu96**
The API route` /transaction/cost` is not working as expected. 
The formula used to compute **gasLimit** for the provided transaction was wrong.

### CODE REVIEW

File: process/smartContract/scQueryService.go [hunk 1]
```diff
@@ -155,9 +156,16 @@ func (service *SCQueryService) ComputeScCallGasLimit(tx *transaction.Transaction
 		return 0, err
 	}
 
-	gasConsumed := service.economicsFee.MaxGasLimitPerBlock(0) - vmOutput.GasRemaining
+	if vmOutput.ReturnCode != vmcommon.Ok {
+		return 0, errors.New(vmOutput.ReturnMessage)
+	}
+
+	moveBalanceGasLimit := service.economicsFee.ComputeGasLimit(tx)
+	gasConsumedExecution := math.MaxUint64 - vmOutput.GasRemaining
```
Comment (iulianpascalau) [inline process/smartContract/scQueryService.go:164]: i do not like this hardcoded value. Why not using tx.GasLimit ? What happens if we decide that we call the SC query with a value different of maxUint?

File: process/transaction/transactionCostEstimator.go [hunk 2]
```diff
@@ -90,11 +91,13 @@ func (tce *transactionCostEstimator) ComputeTransactionGasLimit(tx *transaction.
 		return tce.computeScCallGasLimit(tx)
 	case process.BuiltInFunctionCall:
 		return tce.computeScCallGasLimit(tx)
+	case process.RelayedTx:
+		// TODO for relayed txs not working yet
```
Comment (iulianpascalau) [inline process/transaction/transactionCostEstimator.go:95]: Maybe rephrase the TODO? Something doesn't sound well :D
Add a Jira task for this?

File: integrationTests/vm/testInitializer.go [hunk 3]
```diff
@@ -794,6 +796,30 @@ func GetVmOutput(gasSchedule map[string]map[string]uint64, accnts state.Accounts
...
+		GetCurrentBlockHeaderCalled: func() data.HeaderHandler {
+			return &block.Header{
+				ShardID: testContext.ShardCoordinator.SelfId(),
+			}
+		},
+	})
+
+	gasLimit, err := scQueryService.ComputeScCallGasLimit(tx)
+	if err != nil {
+		fmt.Println("ERROR at ComputeScCallGasLimit()", err)
```
[the remark this answers is not in the document]
Comment (miiu96) [reply integrationTests/vm/testInitializer.go:816]: done.

File: process/smartContract/scQueryService.go [hunk 4]
```diff
@@ -155,9 +156,16 @@ func (service *SCQueryService) ComputeScCallGasLimit(tx *transaction.Transaction
 		return 0, err
 	}
 
-	gasConsumed := service.economicsFee.MaxGasLimitPerBlock(0) - vmOutput.GasRemaining
+	if vmOutput.ReturnCode != vmcommon.Ok {
+		return 0, errors.New(vmOutput.ReturnMessage)
```
[the remark this answers is not in the document]
Comment (miiu96) [reply process/smartContract/scQueryService.go:160]: I think yes. @sasurobert can you please confirm if is ok

File: process/smartContract/scQueryService.go (hunk 1 above)
Comment (miiu96) [reply process/smartContract/scQueryService.go:164]: refactored this part.

File: process/transaction/transactionCostEstimator.go (hunk 2 above)
Comment (miiu96) [reply process/transaction/transactionCostEstimator.go:95]: EN-9040 was created for this.

File: integrationTests/vm/testInitializer.go [hunk 5]
```diff
@@ -813,7 +813,7 @@ func ComputeGasLimit(gasSchedule map[string]map[string]uint64, testContext *VMTe
 
 	gasLimit, err := scQueryService.ComputeScCallGasLimit(tx)
 	if err != nil {
-		fmt.Println("ERROR at ComputeScCallGasLimit()", err)
+		log.Error("error at ComputeScCallGasLimit()", err)
```
Comment (iulianpascalau) [inline integrationTests/vm/testInitializer.go:816]: log.Error("ComputeScCallGasLimit()", "error", err)

Comment (miiu96) [reply integrationTests/vm/testInitializer.go]: done.

File: process/smartContract/scQueryService.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/smartContract/scQueryService.go:162]: I think you should not return error. but should return tx.GasLimit and nil.

**Review by sasurobert — CHANGES_REQUESTED** [review]
do not return error if vmoutput.returnCode != vmcommon.Ok

**Review by sasurobert — DISMISSED** [review]

File: data/transaction/apiTransactionResult.go [hunk 6]
```diff
@@ -82,3 +82,9 @@ type ReceiptApi struct {
 	Data    string   `json:"data,omitempty"`
 	TxHash  string   `json:"txHash"`
 }
+
+// CostResponse is structure used to return the transaction cost in gas units
+type CostResponse struct {
+	GasUnits   uint64 `json:"txGasUnits"`
```
[the remark this answers is not in the document]
Comment (miiu96) [reply data/transaction/apiTransactionResult.go:88]: done.

File: data/transaction/apiTransactionResult_test.go [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+	}
+
+	costResponseBytes, err := json.Marshal(costResponse)
+	assert.Nil(t, err)
+	assert.NotNil(t, costResponseBytes)
+
+	costResponseMap := make(map[string]interface{})
+	err = json.Unmarshal(costResponseBytes, &costResponseMap)
+
+	// DO NOT CHANGE THIS CONST IN ORDER TO KEEP THE BACKWARDS COMPATIBILITY
```
Comment (iulianpascalau) [inline data/transaction/apiTransactionResult_test.go:26]: could have used lower case. :D

**Review by iulianpascalau — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Review by sasurobert — APPROVED** [review]

**Review by stefanandy — APPROVED** [review]
System Testing Passed
